src/jit/importer.cpp

```diff
@@ -2207,6 +2207,13 @@ bool Compiler::impSpillStackEntry(unsigned level,
     /* Assign the spilled entry to the temp */
     impAssignTempGen(tnum, tree, verCurrentState.esStack[level].seTypeInfo.GetClassHandle(), level);
 
+    // If temp is single def and a ref type, grab what type info we can.
+    if (lvaTable[tnum].lvIsTemp && (lvaTable[tnum].lvType == TYP_REF))
+    {
+        CORINFO_CLASS_HANDLE stkHnd = verCurrentState.esStack[level].seTypeInfo.GetClassHandle();
+        lvaSetClass(tnum, tree, stkHnd);
+    }
```

Comment:
For devirt we can't use `lvIsSingleDef` as currently constituted, for a couple of reasons:
- It is computed from a global tree walk, and so is not available when importing code when we are trying to devirtualize
- It currently does not consider some constructs as "single def" that can be treated that way during type propagation -- for example, question ops

If you're suggesting that we eliminate/deprecate/redefine `lvIsTemp` and start setting `lvIsSingleDef` during the importer (and update subsequent code to verify that if it was set during importation it was set correctly) then perhaps we can use it, if we can reconcile the question ops issue. But that is future work.

I'll update the code here to capture the type for newly introduced temps instead of looking at `lvIsTemp`; this better captures the intent anyways.

As for return temps -- I have not gone and comprehensively tried to capture types for all temps, in part because some (like this one) may have multiple definitions, and we don't have a general way of computing bounds on types when we have multiple definitions (specifically for cases where we have mixtures of exact and shared types).

We use the return temp more often these days, since it's also used to avoid interference from the post-inline unpin/gc ref nulling, and so quite often it may be a single def. So we could get that case and we can always use the type from the method signature as the initial approximation. I'll leave those as future enhancements.